### DISCUSSION TITLE
Is there any way to get gtk_file_dialog_get_file binding working in Windows?

### DISCUSSION THREAD

**Discussion by PlatinumLucario** [gircore/gir.core#884] [Q&A] [answered]
This might be an issue with the C library itself, or it may be that I haven't implemented the binding correctly, but when I placed these in:
```cs
    [DllImport("libgio-2.0.so.0", EntryPoint = "g_file_get_path")]
    private static extern string LinuxGetPath(nint file);

    [DllImport("libgio-2.0.0.dylib", EntryPoint = "g_file_get_path")]
    private static extern string MacOSGetPath(nint file);

    [DllImport("libgio-2.0-0.dll", EntryPoint = "g_file_get_path")]
    private static extern string WindowsGetPath(nint file);
```
And then called them in this:
```cs
    public static string GetPath(nint path)
    {
        if (IsLinux())
        {
            return LinuxGetPath(path);
        }
        else if (IsMacOS())
        {
            return MacOSGetPath(path);
        }
        else if (IsWindows())
        {
            return WindowsGetPath(path);
        }
        return path.ToString();
    }
```
From this:
```cs
                var d = FileDialog.New();
                d.SetTitle(Strings.MenuOpenMP2K);
                var filters = Gio.ListStore.New(FileFilter.GetGType());
                filters.Append(filterGBA);
                filters.Append(allFiles);
                d.SetFilters(filters);
                _openCallback = (source, res, data) =>
                {
                    var fileHandle = d.OpenFinish(res, IntPtr.Zero);
                    if (fileHandle != IntPtr.Zero)
                    {
                        var path = FileDialog.GetPath(fileHandle);
                        OpenMP2KFinish(path);
                        d.Unref();
                    }
                    d.Unref();
                };
                d.Open(Handle, IntPtr.Zero, _openCallback, IntPtr.Zero);
```
And once it reaches the GetPath method and attempts to use the DLLImport of the C library function, here's what the debugger says:
```
Fatal error. System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
   at Interop+Ole32.CoTaskMemFree(IntPtr)
   at Gtk.FileDialog.WindowsGetPath(IntPtr)
   at Gtk.FileDialog.GetPath(IntPtr)
   at Kermalis.VGMusicStudio.GTK4.MainWindow+<>c__DisplayClass141_1.<OpenMP2K>b__1(IntPtr, IntPtr, IntPtr)
   at Gio.Internal.Application.Run(IntPtr, Int32, System.String[])
   at Gio.Internal.Application.Run(IntPtr, Int32, System.String[])
   at Kermalis.VGMusicStudio.GTK4.Program.Run(System.String[])
   at Kermalis.VGMusicStudio.GTK4.Program.Main(System.String[])
```
This memory allocation issue only happens in Windows. It doesn't happen on Linux distributions, because it proceeds past it without any issues and can allocate the memory just fine.

So I'm really curious if this is an issue with the C library itself? Am I missing something OS-specific? Have I not implemented the GetPath method correctly? Is there any workarounds that I need to make?

**Comment by fsobolev** [comment, marked as answer]
I don't know what's causing the issue, but a workaround is to make `g_file_get_path()` return IntPtr and then get string using `Marshal.PtrToStringUTF8` https://github.com/NickvisionApps/Application/blob/windows/NickvisionApplication.GNOME/Views/MainWindow.cs#L225

**Reply by PlatinumLucario** [reply]
Awesome! That definitely worked! Now I can see it as a full file path now, like it normally would, instead of a pointer address. Now it can proceed further.

**Reply by cameronwhite** [reply]
This is roughly the same as #791 which has some explanation of the issue

**Reply by badcel** [reply]
Is this fixed in current `main` already? Otherwise we should open a new issue.

**Reply by cameronwhite** [reply]
I think these manual bindings could now use PlatformStringOwnedHandle (or whatever the appropriate type is), but hopefully these manual bindings aren't needed at all in the latest main?

**Reply by badcel** [reply]
I opened #886 to track this.

@PlatinumLucario You should be able to use `[DllImport("Gio", EntryPoint = "g_file_get_path")]` or `[DllImport(Gio.Internal.ImportResolver.Library, EntryPoint = "g_file_get_path")]` and the system automatically figures out the rest.

You don't need to define platform specific library names, this is what the import resolver does automatically if the binary name matches the library name.

**Reply by PlatinumLucario** [reply]
@badcel Hm? That's odd, by changing it to ``[DllImport(Gio.Internal.ImportResolver.Library, EntryPoint = "g_file_get_path")]`` it results in CS0122. Also, I did try ``[DllImport("Gtk", EntryPoint = "gtk_file_dialog_new")]`` with the Gtk ones, and it gives this exception:
```
System.DllNotFoundException: 'Unable to load DLL 'Gtk' or one of its dependencies: The specified module could not be found. (0x8007007E)'
```
So I'm actually curious if there's something I skipped that's needed for import resolver?

**Reply by badcel** [reply]
Actually I did not try my suggestions.

Perhaps the library name must be lower case like in the sample from @fsobolev?

If you check the value from `Gio.Internal.ImportResolver.Library` it should be a constant string containing the correct name of the api if it is part of the Gio binary.

Or your binary names do not match the ones defined in the gir file? This would be interesting, too. Could you check?

**Reply by fsobolev** [reply]
> Perhaps the library name must be lower case like in the sample from @fsobolev?

I used custom DLLImportResolver https://github.com/NickvisionApps/Application/blob/windows/NickvisionApplication.GNOME/Program.cs#L136

**Reply by PlatinumLucario** [reply]
> I used custom DLLImportResolver https://github.com/NickvisionApps/Application/blob/windows/NickvisionApplication.GNOME/Program.cs#L136

@fsobolev That's an awesome workaround, and it works really well too, and in my case (which just copies the libraries into in the same directory as the executable), it can also be made to not use the environmental variable path too
```cs
private static void ImportNativeLibrary() => NativeLibrary.SetDllImportResolver(Assembly.GetExecutingAssembly(), LibraryImportResolver);
```

**Reply by badcel** [reply]
Ah you need a custom loader because the code is outside of the GirCore binary. Sorry for that!

Of course it can't work as the library loader is only active in gircore DLL`s.

Additionally it is written in a way that even if you activate it in your own binary it could break other native interop calls as it only supports one DLL at a time (e.g. if you call GTK and Gio you would have a conflict).

@fsobolev i don't know how much it is worth but a Microsoft employee confirmed that it is okay to cache the result of `NativeLibrary.Load` to avoid calling into the OS to search for the binary at each call.
